### PR TITLE
Controller-executor-kubernetes integration #56 #20 #18

### LINKED ISSUE #56: story: add bi-directional streaming of jobs and logs between CP and executor
When a new job is to be sent to a free executor, CP should be able to stream the job to the executor
When a job is finished the job-pod logs needs to streamed to the CP 

Expected behaviour
Bi-directional streaming must be integrated for grpc server between CP and executor

### PR DESCRIPTION
**Pull request by djChatt**
Description

This pull request adds the functionalities:
 

- [x] executor to request for job from controller and receive them

- [x] executor to run jobs in Kubernetes cluster

- [x] executor to send job logs back to controller

- [x] controller to save the logs/execution context in etcd

fixes #56 #20 #18 

Present Behaviour

No such behaviour present.

Expected Behaviour

There will be full controller-executor-kubernetes integration

### CODE REVIEW

File: internal/controller/server/executor_services.go [hunk 1]
```diff
@@ -59,3 +59,39 @@ func (e *executorCPServicesServer) Register(ctx context.Context, request *execut
...
+	}
+
+	ctx = context.WithValue(ctx, util.ContextKeyUUID, uuid)
+	log.Info(fmt.Sprintf("request id: %v, recieved execution data: %+v", uuid, executionData))
+	err = e.procExec.SaveJobExecutionData(ctx, executionData)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("request id: %v, executor id: %s, error while saving job execution logs", uuid, executionData.ExecutorID))
+		return &executorCPproto.Acknowledgement{Recieved: true}, err
+	}
+	return &executorCPproto.Acknowledgement{Recieved: true}, nil
```
Comment (JaiminRathod) [inline internal/controller/server/executor_services.go:96]: You can replace nil with err in 96th line and remove 94th line

File: internal/executor/client/client.go [hunk 2]
```diff
@@ -45,3 +48,15 @@ func (g *GrpcClient) Register(request *executorCPproto.RegisterRequest) (*execut
```
Comment (JaiminRathod) [inline internal/executor/client/client.go:48]: remove this empty line

File: internal/executor/daemon/executor.go [hunk 3]
```diff
@@ -57,7 +85,7 @@ func (e *executorClient) StartClient(executorConfig config.OctaviusExecutorConfi
 	}
 	e.kubernetesClient, err = kubernetes.NewKubernetesClient(kubeConfig)
 	if err != nil {
-		return status.Error(codes.Internal, err.Error())
+		return err
 	}
 	return nil
```
Comment (JaiminRathod) [inline internal/executor/daemon/executor.go:90]: You can remove if statement and do direct return

File: internal/executor/daemon/executor_test.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+	v1 "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	log.Init("info", "", false)
+}
+
+func TestStartClient(t *testing.T) {
+
```
Comment (JaiminRathod) [inline internal/executor/daemon/executor_test.go:23]: remove empty test function

**Review by JaiminRathod — CHANGES_REQUESTED** [review]

File: internal/executor/daemon/executor_test.go (hunk 4 above)
Comment (littlestar642) [reply internal/executor/daemon/executor_test.go:30]: written the test

File: internal/controller/server/execution/execution.go [hunk 5]
```diff
@@ -4,11 +4,12 @@ import (
...
 	"fmt"
+	"octavius/internal/pkg/constant"
+
 	executorRepo "octavius/internal/controller/server/repository/executor"
 	jobRepo "octavius/internal/controller/server/repository/job"
 	metadataRepo "octavius/internal/controller/server/repository/metadata"
 	"octavius/internal/controller/server/scheduler"
-	"octavius/internal/pkg/constant"
 	"octavius/internal/pkg/idgen"
 	"octavius/internal/pkg/log"
```
[the remark this answers is not in the document]
Comment (littlestar642) [reply internal/controller/server/execution/execution.go:14]: yeah. sorted

File: internal/controller/server/executor_services.go (hunk 1 above)
Comment (littlestar642) [reply internal/controller/server/executor_services.go:96]: this is not a best practice. As such we have to check error and correspondingly return it

File: internal/controller/server/execution/execution.go [hunk 6]
```diff
@@ -202,6 +205,34 @@ func (e *execution) ExecuteJob(ctx context.Context, executionData *clientCPproto
...
+
+	metadataName := clientJob.JobName
+	metadata, err := e.metadataRepo.GetValue(ctx, metadataName)
+	if err != nil {
+		return nil, err
+	}
+	imageName := metadata.ImageName
+
+	job := &executorCPproto.Job{
+		HasJob:    "yes",
```
Comment (faruqisan) [inline internal/controller/server/execution/execution.go:222]: change HasJob to boolean type, yes/no in a string form is not good

File: internal/executor/daemon/executor.go [hunk 7]
```diff
@@ -91,3 +105,121 @@ func (e *executorClient) StartPing() {
...
 }
+
+func (e *executorClient) StartKubernetesService() {
+	for {
+		job, err := e.FetchJob()
+		if err != nil {
+			log.Fatal(fmt.Sprintf("error in getting job from server, error details: %s", err.Error()))
+			time.Sleep(5 * time.Second)
+		}
+		if job.HasJob == "no" {
```
Comment (faruqisan) [inline internal/executor/daemon/executor.go:116]: change job.HasJob to boolean type... who's idea is using string for this

File: internal/controller/server/execution/execution.go (hunk 6 above)
Comment (littlestar642) [reply internal/controller/server/execution/execution.go:222]: There were some issue with using false when zero values got assumed on nil response. Did handled it now so changing back to boolean.

File: internal/controller/server/execution/execution.go [hunk 8]
```diff
@@ -135,37 +139,43 @@ func startExecutorHealthCheck(e *execution, activeExecutorMap *activeExecutorMap
...
 }
 func (e *execution) UpdateExecutorStatus(ctx context.Context, request *executorCPproto.Ping, pingTimeOut time.Duration) (*executorCPproto.HealthResponse, error) {
 	executorID := request.ID
-	clock := clockwork.NewRealClock()
 	// if executor is already active
 	if executor, ok := e.activeExecutorMap.Get(executorID); ok {
-		executor.healthChan <- request.State
-		executor.timer = clock.After(pingTimeOut)
+		if request.State!="" {
+		executor.statusChan <- request.State
```
Comment (faruqisan) [inline internal/controller/server/execution/execution.go:151]: this looks doesn't formated with go fmt

File: internal/executor/daemon/executor.go [hunk 9]
```diff
@@ -78,16 +89,158 @@ func (e *executorClient) RegisterClient() (bool, error) {
 }
 
 func (e *executorClient) StartPing() {
+	ticker := time.NewTicker(5 * time.Second)
```
Comment (faruqisan) [inline internal/executor/daemon/executor.go:92]: ping ticker should be configurable through config

File: internal/executor/daemon/executor.go [hunk 10]
```diff
@@ -78,16 +89,158 @@ func (e *executorClient) RegisterClient() (bool, error) {
...
+			res, err = e.grpcClient.Ping(&executorCPproto.Ping{ID: e.id, State: e.state})
+			if err != nil {
+				log.Fatal(err.Error())
+				return
+			}
+			if !res.Recieved {
+				log.Error(errors.New("ping not acknowledeged by control plane"), "")
+				return
+			}
+		} else {
```
Comment (faruqisan) [inline internal/executor/daemon/executor.go:175]: you can move this code to the first check condition. so the code more readable.

example:

```go
if !job.HasJob {
  time.Sleep(5 * time.Second)
  continue
}

.
.
.

// the rest of code for condition if HasJob is true
```

File: internal/executor/daemon/executor.go [hunk 11]
```diff
@@ -78,16 +89,158 @@ func (e *executorClient) RegisterClient() (bool, error) {
...
+	return e.grpcClient.SendExecutionContext(jobContext)
+}
+
+func (e *executorClient) startWatch(executionContext *executorCPproto.ExecutionContext) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	log.Info(fmt.Sprintf("start watch process for job With k8s name: %s , job ID: %s, and status: %s", executionContext.JobK8SName, executionContext.JobID, executionContext.Status))
+
+	err := e.kubernetesClient.WaitForReadyJob(ctx, executionContext.JobK8SName, e.kubeLogWaitTime)
+
```
Comment (faruqisan) [inline internal/executor/daemon/executor.go:192]: remove whitespace

File: internal/executor/daemon/executor_test.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	v1 "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	log.Init("info", "", false,1)
```
Comment (faruqisan) [inline internal/executor/daemon/executor_test.go:22]: add space after comma, `log.Init("info", "", false, 1)` do you run go fmt or no?

**Review by littlestar642 — APPROVED** [review]

**Pull request merged**